src/generic/datavgen.cpp

```diff
@@ -631,6 +633,101 @@ class wxDataViewTreeNode
 
 WX_DEFINE_SORTED_ARRAY_SIZE_T(unsigned int, wxDataViewSelection);
 
+
+struct RangeEntity
+{
+    unsigned int from;
+    unsigned int to;
+};
+
+WX_DECLARE_OBJARRAY(RangeEntity, ArrayOfRangeEntities);
+#include "wx/arrimpl.cpp"
+WX_DEFINE_OBJARRAY(ArrayOfRangeEntities)
+
+class RowRange
+{
+public:
+    void Add(const unsigned int idx);
+    void RemoveFrom(const unsigned int idx);
+    bool Has(const unsigned int idx);
+    unsigned int Count();
+    unsigned int CountTo(const unsigned int idx);
+    unsigned int GetSize(); // for debugging statistics
+
+private:
+    ArrayOfRangeEntities m_ranges;
+    void CleanUp(int rngIdx);
+};
+
+WX_DECLARE_HASH_MAP(unsigned int, RowRange*, wxIntegerHash, wxIntegerEqual,
+    HeightToRowRangeMap);
+
+/**
+    @class HeightCache
+
+    HeightCache implements a cache mechanism for the DataViewCtrl to give
+    fast access to:
+     * the height of one line (GetLineHeight)
+     * the y-coordinate where a row starts (GetLineStart)
+     * and vice versa (GetLineAt)
+
+    The layout of the cache is a hashmap where the keys are all exisiting
+    row heights in pixels. The values are RowRange objects that represents
+    all having the specified height.
+
+    {
+      22: RowRange([0..10], [15..17], [20..2000]),
+      42: RowRange([11..12], [18..18]),
+      62: RowRange([13..14], [19..19])
+    }
+
+    Examples
+    ========
+
+    GetLineStart
+    ------------
+    To retrieve the y-coordinate of item 1000 it is neccessary to look into
+    each key of the hashmap *m_heightToRowRange*. Get the row count of
+    indices lower than 1000 (RowRange::CountTo) and multiplies it wich the
+    according height.
+
+    RowRange([0..10], [15..17], [20..2000]).CountTo(1000)
+    --> 0..10 are 11 items, 15..17 are 3 items and 20..1000 are 980 items (1000-20)
+        = 11 + 3 + 980 = 994 items
+
+    GetLineStart(1000) --> (22 * 994) + (42 * 3) + (62 * 3) = 22180
+
+    GetLineHeight
+    -------------
+    To retrieve the line height look into each key and check if row is
+    contained in RowRange (RowRange::Has)
+
+    GetLineAt
+    ---------
+    To retrieve the row that starts at a specific y-coordinate.
+    Look into each key and count all rows.
+    Use bisect algorithm in combination with GetLineStart() to
+    find the appropriate item
+*/
+class HeightCache
+{
+public:
+    bool GetLineStart(unsigned int row, int &start);
```

Comment:
This method, as well as the 2 below it, should also be const.

I also think that it would make more sense to use a simpler signature

    // Return the line starting position, in pixels, or -1 if it's not in the cache.
    int GetLineStart(unsigned row) const;

(again for this and the other 2 ones below too).